// vamtiger-database-schema/index.js
var mongoose = require("mongoose")

class Schema {
    constructor() {
        this.userRegistration = this._userRegistration;
        this.directMessage = this._directMessage;
    }

    /**
     * This will return the schema for User Registration.
     **/
    get _userRegistration() {
        var schema = mongoose.Schema({
            // This will reference the User email.
            email: {
                type: String,
                required: true,
                unique: true
            },
            // This will reference the User password.
            password: {
                type: String,
                required: true
            },
            // This will reference whether or not the user is online.
            online: Boolean,
            // This will reference a boolean indicating whether the User email
            // has been validated.
            email_validation: Boolean,
            // This will reference the User registration date.
            registration_date: Date
        });

        return schema;
    }

    /**
     * This will return the schema for direct messages.
     **/
    get _directMessage() {
        var schema = mongoose.Schema({
            // This will reference the ID of the messsage sender.
            sender: String,
            // This will reference the ID of the messsage recipient.
            recipient: String,
            // This will reference the message.
            message: String,
            // The will reference the date at the which the messsage was sent.
            date: Date
        });

        return schema;
    }
}

module.exports = Schema;
